### PR TITLE
journal: init m_object_bytes when constructing ObjectRecorder

### PR DESCRIPTION
**Pull request by zhscn**
The object might overflow if skip this step.

```
$ rg "m_object_bytes" out/client.admin.*
out/client.admin.74578.log
826:2024-05-13T18:51:00.492+0800 7fbb6b4006c0 20 ObjectRecorder: 0x7fbb4c002770 ObjectRecorder (journal_data.1.10233cbe3df0.0): m_object_bytes 0
852:2024-05-13T18:51:00.493+0800 7fbb6b4006c0 20 ObjectRecorder: 0x7fbb4c004070 ObjectRecorder (journal_data.1.10233cbe3df0.1): m_object_bytes 0
878:2024-05-13T18:51:00.493+0800 7fbb6b4006c0 20 ObjectRecorder: 0x7fbb4c003b30 ObjectRecorder (journal_data.1.10233cbe3df0.2): m_object_bytes 0
904:2024-05-13T18:51:00.493+0800 7fbb6b4006c0 20 ObjectRecorder: 0x7fbb4c003d30 ObjectRecorder (journal_data.1.10233cbe3df0.3): m_object_bytes 0

out/client.admin.74621.log
26441:2024-05-13T18:51:41.246+0800 7f6c44c006c0 20 ObjectRecorder: 0x7f6c280026d0 ObjectRecorder (journal_data.1.10233cbe3df0.0): m_object_bytes 13353657
26466:2024-05-13T18:51:41.247+0800 7f6c44c006c0 20 ObjectRecorder: 0x7f6c28003f70 ObjectRecorder (journal_data.1.10233cbe3df0.1): m_object_bytes 13353657
26491:2024-05-13T18:51:41.248+0800 7f6c44c006c0 20 ObjectRecorder: 0x7f6c28003a90 ObjectRecorder (journal_data.1.10233cbe3df0.2): m_object_bytes 13353600
26516:2024-05-13T18:51:41.248+0800 7f6c44c006c0 20 ObjectRecorder: 0x7f6c28003cc0 ObjectRecorder (journal_data.1.10233cbe3df0.3): m_object_bytes 13353600
369259:2024-05-13T18:51:50.300+0800 7f6c5ca006c0 20 ObjectRecorder: 0x7f6c4c0096c0 ObjectRecorder (journal_data.1.10233cbe3df0.4): m_object_bytes 0
371478:2024-05-13T18:51:50.305+0800 7f6c5ca006c0 20 ObjectRecorder: 0x7f6c280026d0 ObjectRecorder (journal_data.1.10233cbe3df0.5): m_object_bytes 0
373699:2024-05-13T18:51:50.311+0800 7f6c5ca006c0 20 ObjectRecorder: 0x7f6c28003f70 ObjectRecorder (journal_data.1.10233cbe3df0.6): m_object_bytes 0
375932:2024-05-13T18:51:50.324+0800 7f6c5ca006c0 20 ObjectRecorder: 0x7f6c28003a90 ObjectRecorder (journal_data.1.10233cbe3df0.7): m_object_bytes 0

```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/journal/ObjectRecorder.cc [hunk 1]
```diff
@@ -42,7 +42,13 @@ ObjectRecorder::ObjectRecorder(librados::IoCtx &ioctx, std::string_view oid,
   }
   m_compat_mode = require_osd_release < CEPH_RELEASE_OCTOPUS;
 
-  ldout(m_cct, 20) << dendl;
+  r = m_ioctx.stat(m_oid, &m_object_bytes, nullptr);
+  if (r < 0) {
+    ldout(m_cct, 5) << "failed to retrieve the size of object: "
```
Comment (idryomov) [inline src/journal/ObjectRecorder.cc:47]: I'm wondering if this would spam the log due to `ENOENT` constantly being returned for new journal objects.

**Review by idryomov — COMMENTED** [review]
> The object might overflow if skip this step.

Can you describe the impact of that happening?  IIRC the overflow is expected in the sense that the OSD (`journal.guard_append` method before octopus or `journal.append` on octopus and later) would return `EOVERFLOW` which is handled in `ObjectRecorder::handle_append_flushed()` by effectively redoing that append.

> out/client.admin.74578.log
> 826:2024-05-13T18:51:00.492+0800 7fbb6b4006c0 20 ObjectRecorder: 0x7fbb4c002770 ObjectRecorder (journal_data.1.10233cbe3df0.0): m_object_bytes 0
>
> out/client.admin.74621.log
> 26441:2024-05-13T18:51:41.246+0800 7f6c44c006c0 20 ObjectRecorder: 0x7f6c280026d0 ObjectRecorder (journal_data.1.10233cbe3df0.0): m_object_bytes 13353657

How were these logs generated?  Was there a crash (or an equivalent like killing the process) involved?

**Comment by zhscn** [comment]
> How were these logs generated? Was there a crash (or an equivalent like killing the process) involved?

Sorry for the unclear description, these logs are just from the lines in this PR and produced by two `rbd bench` runs. The details described as follows:

Actually the crash occured in crimson at this line:
https://github.com/ceph/ceph/blob/5327afb7f7037747d2660d8eadd502f925befc65/src/crimson/os/seastore/object_data_handler.cc#L1039
Here, `size` refers to the `offset + length` of the request to journal data objects, and `max_object_size` is determined by `seastore_default_max_object_size`, which is `16MiB` by default.

If `m_object_bytes` differs from the actual usage, then `size` is smaller than the correct value. Finally, although ObjectRecorder's record doesn't overflow, the data size stored by the underlying backend overflows.

https://github.com/ceph/ceph/blob/5327afb7f7037747d2660d8eadd502f925befc65/src/journal/ObjectRecorder.cc#L322-L330

----

A simple method to verify overflow:

First make `PrimaryLogPG::do_osd_ops` display oid:
```cpp
    case CEPH_OSD_OP_STAT:
      // note: stat does not require RD
      {
	tracepoint(osd, do_osd_op_pre_stat, soid.oid.name.c_str(), soid.snap.val);

	if (obs.exists && !oi.is_whiteout()) {
	  encode(oi.size, osd_op.outdata);
	  encode(oi.mtime, osd_op.outdata);
	  dout(10) << oi.soid << " stat oi has " << oi.size << " " << oi.mtime << dendl; // this line
	} else {
	  result = -ENOENT;
	  dout(10) << "stat oi object does not exist" << dendl;
	}

	ctx->delta_stats.num_rd++;
      }
      break;
```

then execute `rbd bench` twice:

```
MON=1 MGR=1 OSD=1 MDS=0 RGW=0 ../src/vstart.sh -X --without-dashboard --redirect-output -n
./bin/ceph daemon osd.0 config set debug_osd 25
./bin/ceph osd pool create rbd
./bin/rbd create test_image --size 1G --image-feature exclusive-lock,journaling
./bin/rbd bench --image test_image --io-type write --io-pattern rand --io-total 50M
./bin/rbd bench --image test_image --io-type write --io-pattern rand --io-total 50M
```

here is an overflow example:

```
2024-05-13T23:38:30.457+0800 7f23932006c0 10 osd.0 pg_epoch: 12 pg[1.0( v 12'14600 (12'4500,12'14600] local-lis/les=7/8 n=265 ec=7/7 lis/c=7/7 les/c/f=8/8/0 sis=7) [0] r=0 lpr=7 luod=12'14586 crt=12'14599 lcod 12'14585 mlcod 12'14585 active+clean] 1:ceaab864:::journal_data.1.101bd6830092.0:head stat oi has 17005089 2024-05-13T23:38:29.207260+0800
```
`1:ceaab864:::journal_data.1.101bd6830092.0:head` 17005089 > 16777216

File: src/journal/ObjectRecorder.cc (hunk 1 above)
Comment (zhscn) [reply src/journal/ObjectRecorder.cc]: Update to `r < 0 && r != -ENOENT`

**Comment by idryomov** [comment]
> If `m_object_bytes` differs from the actual usage, then `size` is smaller than the correct value. Finally, although ObjectRecorder's record doesn't overflow, the data size stored by the underlying backend overflows.

I don't think the journal object size was intended to be a hard limit, i.e. one that you could set `seastore_default_max_object_size` or some other object store backend limit to directly.  If you look at `ObjectRecorder` code, the respective field is named `m_soft_max_size` and it's referred to as the "soft object size" elsewhere.

This is also evident from the implementation of OSD class methods that I mentioned above.  The newer `journal.append` checks the current size of the object, without taking the length of the data to be appended into account:
https://github.com/ceph/ceph/blob/7528bf51175c3f339f4585311fe946dccd8e3515/src/cls/journal/cls_journal.cc#L1157-L1176

The older `journal.guard_append` doesn't even have the length of the data to be appended at its disposal.

Adding @trociny to keep me honest.

**Comment by idryomov** [comment]
> I don't think the journal object size was intended to be a hard limit, i.e. one that you could set `seastore_default_max_object_size` or some other object store backend limit to directly.

`make check` failure is related:
```
[ RUN      ] TestObjectRecorder.Overflow
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/journal/test_ObjectRecorder.cc:459: Failure
Value of: object2->append(std::move(append_buffers))
  Actual: true
Expected: false
2024-05-13T17:09:56.394+0000 7fd3fe90cb00  0 
recursive lock of object_recorder_lock_2 (22)
 ceph version Development (no_version) squid (dev)
 1: (ceph::mutex_debug_detail::mutex_debugging_base::_will_lock(bool)+0x41) [0x7fd401809701]
 2: (ceph::mutex_debug_detail::mutex_debug_impl<false>::lock(bool)+0x41) [0x558b9fbfd301]
 3: (std::unique_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >::lock()+0x4e) [0x558b9fc1458e]
 4: (std::unique_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >::unique_lock(ceph::mutex_debug_detail::mutex_debug_impl<false>&)+0x31) [0x558b9fc14491]
 5: (journal::ObjectRecorder::flush(Context*)+0x205) [0x558b9fd8dd15]
 6: (TestObjectRecorder::ObjectRecorderFlusher::~ObjectRecorderFlusher()+0xa8) [0x558b9fcdb308]
 7: (TestObjectRecorder_Overflow_Test::TestBody()+0x15ef) [0x558b9fcda70f]
 8: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x7b) [0x558b9fd2d8cb]
 9: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x6a) [0x558b9fd129da]
 10: (testing::Test::Run()+0xc3) [0x558b9fcf0a73]
 11: (testing::TestInfo::Run()+0xda) [0x558b9fcf17ba]
 12: (testing::TestSuite::Run()+0xfb) [0x558b9fcf1fdb]
 13: (testing::internal::UnitTestImpl::RunAllTests()+0x442) [0x558b9fcff6f2]
 14: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x7b) [0x558b9fd309ab]
 15: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x6a) [0x558b9fd152fa]
 16: (testing::UnitTest::Run()+0xcb) [0x558b9fcff25b]
 17: (RUN_ALL_TESTS()+0x11) [0x558b9fbfcdb1]
 18: main()
 19: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fd3ff4d2d90]
 20: __libc_start_main()
 21: _start()
```

**Comment by zhscn** [comment]
Thanks for the detailed explanation, it looks like I opened an inappropriate PR, I will close it.

**Comment by zhscn** [comment]
Is it reasonable to change the overflow condition to `size + data.length() >= soft_max_size` ?
https://github.com/ceph/ceph/blob/7528bf51175c3f339f4585311fe946dccd8e3515/src/cls/journal/cls_journal.cc#L1164-L1168

**Comment by idryomov** [comment]
> Is it reasonable to change the overflow condition to `size + data.length() >= soft_max_size` ?

Without a strong reason, probably not -- journaling has behaved this way for 8 years.

**Pull request closed without merging**